### PR TITLE
JS: Add steps into static regexp capture group references

### PR DESCRIPTION
**Pull request by asgerf**
Adds taint steps from regexp matching to a following use of `RegExp.$1` or similar.

- This [query console](https://lgtm.com/query/959605277707050381/) run shows a bunch of new taint steps.
- [security/smoke-test](https://git.semmle.com/asger/dist-compare-reports/tree/js/static-regexp-capture-group-step_1593209978379) looks uneventful, but also showing a couple of new taint steps.

Note that we don't check if the regexp in the capture group is restrictive enough to sanitize the input. I've (re)created an issue for it [here](https://github.com/github/codeql-javascript-team/issues/198).

Also note that `RegExp.$1` and friends are only assigned to on a _successful_ match. If the match fails, the old values remain. I added test cases to explain this, but couldn't find any real-world examples where this mattered, so I kept it simple by just assuming every match erases the old values, regardless of whether there is a success check.

Fixes https://github.com/github/codeql/issues/3739.

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/dataflow/TaintTracking.qll [hunk 1]
```diff
@@ -705,6 +705,75 @@ module TaintTracking {
...
+     * a static property of `RegExp`.
+     */
+    private predicate staticRegExpCaptureStep(DataFlow::Node pred, DataFlow::Node succ) {
+      getACaptureSetter(pred) = getANodeReachingCaptureRef(succ)
+      or
+      exists(DataFlow::MethodCallNode replace |
+        replace.getMethodName() = "replace" and
+        getANodeReachingCaptureRef(succ) = replace.getCallback(1).getFunction().getEntry() and
+        pred = replace.getReceiver()
+      )
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/dataflow/TaintTracking.qll:764]:
```suggestion
      exists(StringReplaceCall replace |
        getANodeReachingCaptureRef(succ) = replace.getRawReplacement().getFunction().getEntry() and
        pred = replace.getReceiver()
      )
```

Comment (asgerf) [reply javascript/ql/src/semmle/javascript/dataflow/TaintTracking.qll]: We'd need `getALocalSource().(DataFlow::FunctionNode).` in there as well, which was a bit verbose, so I added `StringReplaceCall.getReplacementCallback()` instead, PTAL.

**Comment by asgerf** [comment]
Running an evaluation on big-apps.

**Comment by asgerf** [comment]
[Big-apps evaluation](https://git.semmle.com/asger/dist-compare-reports/tree/js/static-regexp-capture-group-step_1593460756379) went fine.

**Comment by asgerf** [comment]
@esbena can I get an approve? You're still requesting changes.

**Pull request merged**